### DISCUSSION TITLE
아이템 16. public 클래스에서는 public 필드가 아닌 접근자 메서드를 사용하라

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#39] [3. 과제] [answered]
> 🤔 왜 `public` 클래스에서는 `public` 필드가 아닌 접근자 메서드를 사용해야 할까?

> [객체지향의 사실과 오해](http://www.yes24.com/Product/Goods/18249021)에서는 이렇게 객체지향의 본질을 정의한다.
> - 객체지향이란 시스템을 상호작용하는 자율적인 객체들의 공동체로 바라보고 객체를 이용해 시스템을 분할하는 방법이다.
> - 자율적인 객체란 `상태(state)`와 `행위(behavior)`를 함께 지니며 스스로 자기 자신을 책임지는 객체를 의미한다.
> - 객체는 시스템의 행위를 구현하기 위해 다른 객체와 협력한다.
> - 각 객체는 협력내에서 정해진 역할을 수행하며 역할은 관련된 책임의 집합이다.
> - 객체는 다른 객체와 협력하기 위해 메시지를 전송하고, 메시지를 수신한 객체는 메시지를 처리하는데 적합한 메서드를 자율적으로 선택한다.

### 자율적인 객체를 보장해 줄 수 있는것은 바로 `캡슐화` 이다.
  - 캡슐화는 객체의 상태, 메서드를 하나로 묶고, 실제 구현 내용을 외부에 감추는 것이다.
  - 외부 객체는 객체 내부의 구조를 얻지 못하며 객체가 노출해서 제공하는 필드와 메서드만 사용 가능하다.
  - 캡슐화의 이유는 객체에게 자율성을 보장하기 위함이다.
  - 자바는 이를 위해 `접근 제어자(Access Modifier)`를 사용하고 있다.

## 예제 1.
``` java
public class DingCook { // 해당 클래스는 인스턴스 필드만 모아놓는 일 외에는 아무 목적이 없다.
  public double height;
  public double weight;
}

class Main {

  public static void main(String[] args) {
    DingCook dingCook = new DingCook();
    dingCook.height = 192.0d;
    System.out.println(dingCook.height + "cm");
    dingCook.weight = 75.5d;
    System.out.println(dingCook.weight + "kg");
  }
}
```
![image](https://user-images.githubusercontent.com/53285909/210128864-4c512c4f-8e01-4d2a-aab9-1abf90fffc2a.png)

- 해당 예제는 위에서 말하고 있는 `캡슐화`의 이점이 제공되지 못한 코드이다.

## 예제 2.
```java
public class DingCook {
    private double height;
    private double weight;
}

class Main {

    public static void main(String[] args) {
        DingCook dingCook = new DingCook();
        dingCook.height = 192.0d;
        System.out.println(dingCook.height + "cm");
        dingCook.weight = 75.5d;
        System.out.println(dingCook.weight + "kg");
    }
}
```
- 자 그럼 객체를 캡슐화 하기 위해 각 필드의 접근 제어자를 `private` 로 은닉화 해보자.
![image](https://user-images.githubusercontent.com/53285909/210129167-81738f2e-7117-4de0-b0b4-70a0ab6c3239.png)
- 인텔리제이는 울부 짖는다. 해당 필드는 `private` 접근 제어자라고 접근 할 수 없다고...
- 자 이제 필드에 직접적으로 접근할 수 있는 것은 객체 본인 뿐이다.
- 예제 2. 에서 Main 클래스가 원했던 것은 무엇이였을까??
  - 1. 키와 몸무게를 세팅
  - 2. 키와 몸무게를 출력
- 한번 생각 해보면 이제 우리는 예제 1과는 달리 직접 키와 몸무게에 접근할 수도 그 값을 바꿀 수가 없다.
- 객체에게 자율성을 부여해 키와 몸무게를 객체 본인이 세팅하게 하는 방법은 무엇일까?
- 그리고 그 객체가 만들어 낸 키와 몸무게를 가져와 출력할 수 있는 방법은 무엇일까?

```java
public class DingCook {
    private double height;
    private double weight;

    public DingCook(double height, double weight) {
        this.height = height;
        this.weight = weight;
    }

    public double getHeight() {
        return height;
    }

    public double getWeight() {
        return weight;
    }

    public void setHeight(double height) {
        this.height = height;
    }

    public void setWeight(double weight) {
        this.weight = weight;
    }
}

class Main {

    public static void main(String[] args) {
        DingCook dingCook = new DingCook(192.0, 75.5);

        System.out.println(dingCook.getHeight() + "cm");
        System.out.println(dingCook.getWeight() + "kg");

        dingCook.setHeight(194.0);
        dingCook.setWeight(100.0);
        System.out.println(dingCook.getHeight() + "cm");
        System.out.println(dingCook.getWeight() + "kg");
    }
}

```
![image](https://user-images.githubusercontent.com/53285909/210129784-02f26940-a394-490f-b73b-6e31bcb00685.png)
- 예제 1. 과 달리 예제 2. 는 객체 스스로가 행동(메서드)을 통해 객체 자신의 상태를 알려준다.

# 📚 최종 정리
- 객체의 필드를 직접 접근해서 가공하는 행위는 개발자의 실수가 유발 될 수있고, 다른 결과를 초례할 수 있다. 그리고 객체가 가진 자율성을 억업한다.
- public 클래스에서는 public 필드를 지양하고 접근자 메서드의 사용을 지향하자
- c.f) package-private 클래스와 private 중첩 클래스는 데이터 필드가 노출 되어도 상관없다.
- [package-private 클래스와 private 중첩 클래스](https://github.com/orgs/Study-2-Effective-Java/discussions/41)

**Comment by corock** [comment, marked as answer]
객체지향의 사실과 오해 내용까지 참고하는 모습 보기 좋습니다 👍
해당 아이템 내용은 객체지향 프로그래밍에서 근간이 되는 특징 중 하나로 봅니다.

`public` 필드 지양은 당연한 얘기고, 접근자 메서드 사용에 대해서 저는 다음 3단계로 생각이 확장되는 것 같네요.

1. 접근 지정자 사용을 지향한다.
2. `setter` 를 **지양**한다.
3. 무조건 `getter` 사용도 **지양**한다.

2번과 3번에 대해 읽을거리가 있으니 한번 읽어보는 것을 추천합니다.

2. [Setter를 지양해야 하는 이유](https://min103ju.github.io/java/no-setter/)
3. [getter를 사용하는 대신 객체에 메시지를 보내자](https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/)

**Reply by Irisation23** [reply]
![image](https://user-images.githubusercontent.com/53285909/210198197-606b9888-4f2a-4e92-92dd-f5e29f04843a.png)
> 출처 - https://tecoble.techcourse.co.kr/post/2020-04-28-ask-instead-of-getter/

링크로 남겨주신 글 다 읽어봤는데 좋은 정리라고 생각됩니다. 그리고 3.의 링크 걸린 글 중 위의 정리가 굉장히 마음에 와 닿네요. 
객체 지향적 코드를 작성 함에 있어서 잃지 말아야 할 개념중 하나는 객체와 객체는 메시지를 주고받는 상호 관계에 있단 사실을 잃지 않아야 겠다는 생각이 듭니다. 그리고 이번 주제를 정리해 보면 타당한 이유없는 코드의 작성을 줄여야 한다는 생각 또한 하게 되었습니다.
그 정리를 위의 링크를 보며 다시금 마음을 다져봅니다.
감사합니다. @corock

**Reply by JoisFe** [reply]
객체 지향의 사실과 오해 
뭔가 얇고 쉬워 보여서 가볍게 읽으려 봤다가
한페이지 한페이지 참 읽는게 어렵더라구요 마치 철학책 같은 느낌... (물론 철학책 한번도 안 읽어 봤습니다 ㅎㅎ)

아무튼 매우 추천!!!

**Comment by JoisFe** [comment]
정리 감사합니다.

public 필드가 아닌 접근자 메서드를 이용해 필드에 접근하는 것을 지향하라는 주제인 것 같습니다.
접근자 메서드를 이용하는 이유가 해당 객체에 대한 필드를 다른 객체에서 직접 접근하고 수정하는 등의 작업은 객체 지향적인 프로그래밍에는 맞지 않기 때문인 것 같습니다. 
다른 객체에서 해당 객체의 필드에 대한 작업을 해야되는 상황이라면 접근자 메서드를 이용 즉 메시지를 통해 해당 객체에게 역할을 주게하는 것이 맞다고 생각하고 이러한 부분이 @bunsung92 님 께서 정리하신
- 객체는 다른 객체와 협력하기 위해 메시지를 전송하고, 메시지를 수신한 객체는 메시지를 처리하는데 적합한 메서드를 자율적으로 선택한다.

에 부합한다고 생각합니다.

그리고 접근 지정자 활용은 #41 를 참고하시면 될 것 같습니다.

**Reply by Irisation23** [reply]
객체에게 자유를 👍

**Comment by coalong** [comment]
정리 감사합니다! 
왜 사용해야 하는지에 대해서 설명을  깔끔하게 잘 써주서 정리가 잘 되었어요~ 👍 

@corock 님이 올려주신 참고자료도 잘 읽었습니다.!

**Comment by chikeem90** [comment]
> 인텔리제이는 울부 짖는다. 해당 필드는 private 접근 제어자라고 접근 할 수 없다고...

기억 남는 한 줄 👍
